libpod/runtime_img.go

```diff
@@ -95,6 +98,260 @@ type imageDecomposeStruct struct {
 	transport   string
 }
 
+func (k *KpodImage) assembleFqName() string {
+	return fmt.Sprintf("%s/%s:%s", k.Registry, k.ImageName, k.Tag)
+}
+
+func (k *KpodImage) assembleFqNameTransport() string {
+	return fmt.Sprintf("%s%s/%s:%s", k.Transport, k.Registry, k.ImageName, k.Tag)
+}
+
+//KpodImage describes basic attributes of an image
+type KpodImage struct {
+	Name          string
+	ID            string
+	fqname        string
+	hasImageLocal bool
+	Runtime
+	Registry       string
+	ImageName      string
+	Tag            string
+	HasRegistry    bool
+	Transport      string
+	beenDecomposed bool
+	PullName       string
+}
+
+// NewImage creates a new image object based on its name
+func (r *Runtime) NewImage(name string) KpodImage {
+	return KpodImage{
+		Name:    name,
+		Runtime: *r,
+	}
+}
+
+// GetImageID returns the image ID of the image
+func (k *KpodImage) GetImageID() (string, error) {
+	if k.ID != "" {
+		return k.ID, nil
+	}
+	image, _ := k.GetFQName()
+	img, err := k.Runtime.GetImage(image)
+	if err != nil {
+		return "", err
+	}
+	return img.ID, nil
+}
+
+// GetFQName returns the fully qualified image name if it can be determined
+func (k *KpodImage) GetFQName() (string, error) {
+	// Check if the fqname has already been found
+	if k.fqname != "" {
+		return k.fqname, nil
+	}
+	err := k.Decompose()
+	if err != nil {
+		return "", err
+	}
+	k.fqname = k.assembleFqName()
+	return k.fqname, nil
+}
+
+func (k *KpodImage) findImageOnRegistry() error {
+	searchRegistries, err := GetRegistries()
+
+	if err != nil {
+		return errors.Wrapf(err, " the image name '%s' is incomplete.", k.Name)
+	}
+
+	for _, searchRegistry := range searchRegistries {
+		k.Registry = searchRegistry
+		err = k.GetManifest()
+		if err == nil {
+			k.fqname = k.assembleFqName()
+			return nil
+
+		}
+	}
+	return errors.Errorf("unable to find image on any configured registries")
+
+}
+
+// GetManifest tries to GET an images manifest, returns nil on success and err on failure
+func (k *KpodImage) GetManifest() error {
+	pullRef, err := alltransports.ParseImageName(k.assembleFqNameTransport())
+	if err != nil {
+		return errors.Errorf("unable to parse '%s'", k.assembleFqName())
+	}
+	imageSource, err := pullRef.NewImageSource(nil)
+	if err != nil {
+		return errors.Errorf("unable to create new image source")
+	}
+	_, _, err = imageSource.GetManifest()
+	if err == nil {
+		return nil
+	}
+	return err
+}
+
+//Decompose breaks up an image name into its parts
+func (k *KpodImage) Decompose() error {
+	k.beenDecomposed = true
```

Comment:
Should we do a beenDecomposed check before this? And just return if the image is already decomposed?

Reply:
good idea